### PR TITLE
Extract streamNdjson middleware and use it in observeDomainEvents api.

### PR DESCRIPTION
**Pull request by yeldiRium**
Hi @goloroden,

as discussed in #339 I've extracted a middleware from the `observeDomainEvents` api that sets up a connection as using `application/x-ndjson` and sends a heartbeat.

I've added tests for the middleware and had to make slight changes to the helper `WriteStream` `asJsonStream` which luckily don't affect the other places in which the helper is used.

Feel free to merge the changes if everything is well.

### CODE REVIEW

File: lib/apis/middlewares/streamNdjson.ts [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+const logger = flaschenpost.getLogger();
+
+const heartbeat = { name: 'heartbeat' };
+
+/**
+ * The streamNdjson middleware initializes a long-lived connection with the
+ * content type `application/x-ndjson` and sends a periodic heartbeat every
+ * `heartbeatInterval` milliseconds.
+ */
```
Comment (goloroden) [inline lib/apis/middlewares/streamNdjson.ts]: Can we change this to a comment using multiple `//`s? I know that we don't enforce this right now, but this is the common style used everywhere else (if we have comments), so I'd vote to stick with that.

File: lib/apis/middlewares/streamNdjson.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  return async function (req, res, next): Promise<void> {
+    try {
+      let heartbeatIntervalId: NodeJS.Timeout;
+
+      req.setTimeout(0, (): void => undefined);
+      res.setTimeout(0);
+
+      res.writeHead(200, { 'content-type': 'application/x-ndjson' });
+
+      res.connection.once('close', (): void => {
```
Comment (goloroden) [inline lib/apis/middlewares/streamNdjson.ts]: Is this event enough to clear things, or should we also run this code in case of errors (see the `catch` below)?

File: lib/apis/observeDomainEvents/http/v2/getDomainEvents.ts [hunk 3]
```diff
@@ -68,27 +62,12 @@ const getDomainEvents = function ({
         /* eslint-enable @typescript-eslint/no-floating-promises */
       };
 
-      req.setTimeout(0, (): void => undefined);
-      res.setTimeout(0);
-
-      res.writeHead(200, { 'content-type': 'application/x-ndjson' });
-
       res.connection.once('close', (): void => {
```
Comment (goloroden) [inline lib/apis/observeDomainEvents/http/v2/getDomainEvents.ts]: Same question here as in the middleware: Is this event enough?

File: test/shared/http/asJsonStream.ts [hunk 4]
```diff
@@ -7,8 +7,9 @@ const asJsonStream = function <TItem> (...handleJson: ((item: TItem) => void)[])
     write (chunk: object, _, callback: (error?: any) => void): void {
       const data = JSON.parse(chunk.toString());
 
-      if (counter > handleJson.length) {
-        return callback(new Error(`Received ${counter + 1} items (${JSON.stringify(data)}), but only expected ${handleJson.length}.`));
+      // Ignore any messages after all handlers have been applied.
+      if (counter >= handleJson.length) {
+        return;
```
Comment (goloroden) [inline test/shared/http/asJsonStream.ts]: Hmmmm 🤔 

Doesn't this stream now stay open forever, because we never close it? Is this by intention?

Oh, and: Why the change from `>` to `>=`?

File: test/unit/apis/middlewares/streamNdjsonTests.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+suite('streamNdjson middleware', (): void => {
+  let api: Application;
+
+  setup(async (): Promise<void> => {
+    api = express();
+    api.get('/', streamNdjsonMiddleware({ heartbeatInterval: 1000 }));
+  });
+
+  test('returns 200.', async (): Promise<void> => {
```
Comment (goloroden) [inline test/unit/apis/middlewares/streamNdjsonTests.ts]: Maybe `returns the status code 200.`?

File: test/unit/apis/middlewares/streamNdjsonTests.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            }
+          ));
+        /* eslint-enable @typescript-eslint/no-floating-promises */
+      } catch (ex) {
+        reject(ex);
+      }
+    });
+  });
+
+  test('returns application/json.', async (): Promise<void> => {
```
Comment (goloroden) [inline test/unit/apis/middlewares/streamNdjsonTests.ts]: This is wrong, isn't it? Shouldn't this be `application/x-ndjson`?

File: test/shared/http/asJsonStream.ts (hunk 4 above)
Comment (yeldiRium) [reply test/shared/http/asJsonStream.ts]: The change from `>` to `>=` because `>` was wrong and did not produce errors in cases where it previously should.

That the stream now stays open forever is intended on my part, so that the helper can accept more stream elements without throwing an exception. The producing site should stop the producing stream. This write stream will probably stay open and produce a memory leak, but since it is only used in tests that shouldn't be a problem.

File: lib/apis/middlewares/streamNdjson.ts (hunk 2 above)
Comment (yeldiRium) [reply lib/apis/middlewares/streamNdjson.ts]: This is enough to clear things, since the interval is created in the last statement in this block and any error occuring there will result in the interval not being set. In that case it doesn't need to be cleared. Errors inside the interval can't be handled here anyway.

File: lib/apis/observeDomainEvents/http/v2/getDomainEvents.ts (hunk 3 above)
Comment (yeldiRium) [reply lib/apis/observeDomainEvents/http/v2/getDomainEvents.ts]: The same argument from the middleware works here. The statement that attaches a listener to the event emitter is the last one in the scope and any error occuring there will result in the cleanup not being necessary.

File: test/shared/http/asJsonStream.ts (hunk 4 above)
Comment (goloroden) [reply test/shared/http/asJsonStream.ts]: Okay, makes sense 👍

File: test/unit/apis/middlewares/streamNdjsonTests.ts (hunk 6 above)
Comment (yeldiRium) [reply test/unit/apis/middlewares/streamNdjsonTests.ts]: True. I've changed it and the test still passes. Something feels wrong here. Why did the test pass in the first place? Supertest doesn't seem to work that great with streams unfortunately.

Do you have any idea on how to improve this?

Comment (goloroden) [reply test/unit/apis/middlewares/streamNdjsonTests.ts]: Hmmmm … good point, and yes, good question 🤔
